tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,344 @@
+from tools.codegen.utils import mapMaybe, Target, assert_never
+
+from tools.autograd.gen_inplace_or_view_type import (
+    gen_formals, modifies_arguments
+)
+from tools.autograd.gen_trace_type import (
+    type_wrapper_name
+)
+
+def is_view_op(f: NativeFunction) -> bool:
+    rets = f.func.returns
+    is_non_mutating_view = len(rets) > 0 and any(r.annotation is not None and not r.annotation.is_write for r in rets)
+    is_inplace_view = f.tag is not None and f.tag is Tag.inplace_view
+    return is_non_mutating_view or is_inplace_view
+
+def return_names_str(f: NativeFunction) -> str:
+    if len(f.func.arguments.out) != 0:
+        if len(f.func.arguments.out) > 1:
+            return_names = ', '.join(a.name for a in f.func.arguments.out)
+            return f'{DispatcherSignature.from_schema(f.func).returns_type().cpp_type()}({return_names});'
+        else:
+            return f.func.arguments.out[0].name
+    if f.func.arguments.self_arg is not None:
+        return f.func.arguments.self_arg.argument.name
+    raise AssertionError("Unable to handle functionalization for op={str(f.func.name)}")
+
+def is_multi_output(rets: Tuple[Return, ...]) -> bool:
+    return len(rets) > 1 or rets[0].type.is_list_like() is not None
+
+def arg_name(a: Argument) -> str:
+    return f'{a.name}_' if a.type.is_tensor_like() else a.name
+
+def unwrap_tensor_args_str(f: NativeFunction) -> str:
+    return '\n      '.join(
+        f'auto {arg_name(arg)} = at::functionalization::impl::from_functional_tensor({arg.name});'
+        for arg in f.func.arguments.flat_all if arg.type.is_tensor_like())
+
+# Generates the Functionalization kernel for:
+# - ops that create aliases (e.g. transpose())
+# - ops that are views AND mutations (e.g. transpose_())
+def emit_view_functionalization_body(f: NativeFunction, g: Optional[NativeFunctionsGroup], native_api_name: str) -> str:
+    dispatcher_sig = DispatcherSignature.from_schema(f.func)
+    dispatcher_exprs = dispatcher_sig.exprs()
+
+    keyset = 'ks & c10::after_func_keyset'
+    return_type = dispatcher_sig.returns_type().remove_const_ref().cpp_type()
+
+    api_name = CppSignatureGroup.from_native_function(
+        f, method=False, fallback_binding=f.manual_cpp_binding).most_faithful_signature().name()
+
+    # view op case
+    assert is_view_op(f)
+
+    if f.tag is not None and f.tag is Tag.inplace_view:
+        # This op is both an inplace op AND a view op.
+        # See Note [Functionalization Pass - Inplace View Ops] for details.
+        # I currently have the view meta call into the out-of-place variant of the view, to avoid
+        # having to define an extra ~20 inplace {view}_inverse_ functions.
+        # Most view ops don't have NativeFunctionGroup's both, because we don't define out= variants for view ops.
+        # For now, I'm dangerously assuming that every inplace-view op has a corresponding out-of-place view op,
+        # with the same name but the trailing underscore removed.
```

Comment:
if you wanna live dangerously, add more asserts (e.g., that the last character is underscore, that the truncated name exists)